drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,435 @@
+#endif /* CONFIG_ADC_ASYNC */
+};
+
+static int adc_gd32_init(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+	int ret;
+
+	data->dev = dev;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+	/* Select adc clock source and its prescaler. */
+	rcu_adc_clock_config(cfg->rcu_clock_source);
+#endif
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+#if defined(CONFIG_SOC_SERIES_GD32F403) || \
+	defined(CONFIG_SOC_SERIES_GD32VF103) || \
+	defined(CONFIG_SOC_SERIES_GD32F3X0)
+	/* Set SWRCST as the regular channel external trigger. */
+	ADC_CTL1(cfg->reg) &= ~ADC_CTL1_ETSRC;
+	ADC_CTL1(cfg->reg) |= CTL1_ETSRC(7);
+
+	/* Enable external trigger for regular channel. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ETERC;
+#endif
+
+	/* Enable ADC */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ADCON;
+
+	adc_gd32_calibration(cfg);
+
+	cfg->irq_config_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+#define HANDLE_SHARED_IRQ(n, active_irq)							\
+	static const struct device *dev_##n = DEVICE_DT_INST_GET(n);				\
+	const struct adc_gd32_config *cfg_##n = dev_##n->config;				\
+												\
+	if ((cfg_##n->irq_num == active_irq) &&							\
+		(ADC_CTL0(cfg_##n->reg) & ADC_CTL0_EOCIE)) {					\
+		adc_gd32_isr(dev_##n);								\
+	}
+
+static void adc_gd32_global_irq_handler(const struct device *dev)
+{
+	const struct adc_gd32_config *cfg = dev->config;
+
+	LOG_DBG("global irq handler: %u", cfg->irq_num);
+
+	DT_INST_FOREACH_STATUS_OKAY_VARGS(HANDLE_SHARED_IRQ, (cfg->irq_num));
+}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+#define ADC_CLOCK_SOURCE(n)									\
+	.rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock)
+#else
+#define ADC_CLOCK_SOURCE(n)
+#endif
+
+#define DT_ADC2 0U
+
+#define ADC_GD32_INIT(n)									\
+	PINCTRL_DT_INST_DEFINE(n);								\
+	static void adc_gd32_irq_cfg_##n(void)							\
+	{											\
+		bool initialized = false;							\
+												\
+		for (size_t i = 0; i < sizeof(adc_gd32_init_irqs); i++) {			\
+			if (adc_gd32_init_irqs[i] == 0U) {					\
+				initialized = true;						\
+				adc_gd32_init_irqs[i] = DT_INST_IRQN(n);			\
+				break;								\
+			}									\
+												\
+			if (adc_gd32_init_irqs[i] == DT_INST_IRQN(n)) {				\
+				LOG_INF("interrupt initialize success, num: %u",		\
+					DT_INST_IRQN(n));					\
+				return;								\
+			}									\
+		}										\
+												\
+		if (initialized) {								\
+			IRQ_CONNECT(DT_INST_IRQN(n),						\
+				DT_INST_IRQ(n, priority),					\
+				adc_gd32_global_irq_handler,					\
+				DEVICE_DT_INST_GET(n),						\
+				0);								\
+			irq_enable(DT_INST_IRQN(n));						\
+			LOG_INF("interrupt initialize success, num: %u, priority: %u",		\
+				DT_INST_IRQN(n), DT_INST_IRQ(n, priority));			\
+		} else {									\
+			LOG_INF("interrupt initialize failed");					\
+		}										\
+	}											\
```

Comment:
I'd personally do the following instead:

1. create a "connect" function that just has `IRQ_CONNECT()` for every instance
2. Move all the initialization logic to the init/sub-init function. If an IRQ hasn't been initialized, just run the "connect" call for that instance. 

Maybe I'm missing something, but I think it's doable.

Reply:
Looks `IRQ_CONNECT` have a `build time` restriction, refer from [aarch32 irq definition](https://github.com/zephyrproject-rtos/zephyr/blob/2ef3fe7bca137dcbf6bad3b13a038d1998aa1a37/include/zephyr/arch/arm/aarch32/irq.h#L115).

Reply:
So, this solution still have problem. It should provide an unique `IRQ_CONNECT` for each interrupt.

Reply:
Config gd32 adc irq through node-label, based on https://github.com/zephyrproject-rtos/zephyr/pull/42215#discussion_r873123172